dali/pipeline/operator/sequence_shape.h

```diff
@@ -0,0 +1,307 @@
+    return range_[idx_];
+  }
+
+  bool operator==(const RangeIterator &other) const {
+    return idx_ == other.idx_;
+  }
+
+  bool operator!=(const RangeIterator &other) const {
+    return !(*this == other);
+  }
+
+  void operator++() {
+    ++idx_;
+  }
+
+ private:
+  const Range &range_;
+  IndexType idx_;
+};
+
+struct SliceView {
+  uint8_t *ptr;
+  TensorShape<> shape;
+  size_t type_size;
+};
+
+
+/**
+ * @brief Iterable range over slices of tensor, created by expansion of the outermost
+ * extents of the tensor. For instance passing ``SliceView`` of shape ``{6, 5, 3, 4}``
+ * and ``ndims_to_unfold=2`` will result in range of 30 ``SliceViews`` of shape ``{3, 4}`` each.
+ */
+class UnfoldedSliceRange {
+ public:
+  using IndexType = size_t;
+  using ValueType = SliceView;
+
+  UnfoldedSliceRange(ValueType view, int ndims_to_unfold)
+      : view_{std::move(view)},
+        num_slices_{[&]() {
+          assert(view_.shape.sample_dim() >= ndims_to_unfold);
+          auto vol = volume(view_.shape.begin(), view_.shape.begin() + ndims_to_unfold);
+          return static_cast<std::make_unsigned_t<decltype(vol)>>(vol);
+        }()},
+        slice_shape_{view_.shape.begin() + ndims_to_unfold, view_.shape.end()},
+        slice_stride_{view_.type_size * volume(slice_shape_)} {}
+
+  RangeIterator<UnfoldedSliceRange> begin() const {
+    return {*this, 0};
+  }
+
+  RangeIterator<UnfoldedSliceRange> end() const {
+    return {*this, NumSlices()};
+  }
+
+  ValueType operator[](IndexType idx) const {
+    return {view_.ptr + idx * SliceSize(), SliceShape(), view_.type_size};
+  }
+
+  size_t NumSlices() const {
+    return num_slices_;
+  }
+
+  TensorShape<> SliceShape() const {
+    return slice_shape_;
+  }
+
+  size_t SliceSize() const {
+    return slice_stride_;
+  }
+
+ private:
+  ValueType view_;
+  size_t num_slices_;
+  TensorShape<> slice_shape_;
+  size_t slice_stride_;
+};
+
+template <typename Backend>
+struct TensorVectorBuilder {
+  TensorVectorBuilder(int num_samples, DALIDataType type, int sample_dim, TensorLayout layout,
+                      bool is_pinned, AccessOrder order)
+      : tv_(num_samples) {
+    tv_.set_type(type);
+    tv_.set_sample_dim(sample_dim);
+    tv_.SetLayout(layout);
+    tv_.set_pinned(is_pinned);
+    tv_.set_order(order);
+  }
+
+  void push(const SliceView &view) {
+    std::shared_ptr<void> ptr(view.ptr, [](void *) {});  // no deleter
+    tv_.UnsafeSetSample(size++, ptr, view.type_size * volume(view.shape), tv_.is_pinned(),
+                        view.shape, tv_.type(), tv_.order(), tv_.GetLayout());
+  }
+
+  TensorVector<Backend> take() {
+    assert(size == tv_.num_samples());
+    return std::move(tv_);
+  }
+
+  TensorVector<Backend> tv_;
+  size_t size = 0;
+};
+
+
+template <typename Backend>
+TensorVectorBuilder<Backend> tv_builder_like(const TensorVector<Backend> &data, int num_samples,
+                                             int ndims_to_unfold = 0) {
```

Comment:
You'd probably hate me for this comment, but I feel that the functions below deserve some tests. Especially those, which operate on `TensorListShape`...